### PR TITLE
Start on documentation

### PR DESCRIPTION
**Pull request by dwhenry**
## Description of change
Start on documentation

## Link to relevant ticket
https://trello.com/c/uQXPrB6U/313-document-the-engine-msf

## Processed Pages:

https://github.com/ministryofjustice/laa-claim-non-standard-magistrate-fee-backend/blob/Q1-S2/docs/gems/laa_multi_step_forms/docs/Setup.md
https://github.com/ministryofjustice/laa-claim-non-standard-magistrate-fee-backend/blob/Q1-S2/docs/gems/laa_multi_step_forms/docs/Steps.md
https://github.com/ministryofjustice/laa-claim-non-standard-magistrate-fee-backend/blob/Q1-S2/docs/gems/laa_multi_step_forms/docs/TaskList.md
https://github.com/ministryofjustice/laa-claim-non-standard-magistrate-fee-backend/blob/Q1-S2/docs/gems/laa_multi_step_forms/docs/Validations.md

### CODE REVIEW

File: gems/laa_multi_step_forms/docs/TaskList.md [hunk 1]
```diff
@@ -1 +1,87 @@
...
+
+example
+
+```ruby
+def completed?
+  Steps::ComplexStep.new(application:).valid?
+end
+```
+
+### #filfilled?
```
Comment (willmcb) [inline gems/laa_multi_step_forms/docs/TaskList.md:77]: Typo here 👍

File: gems/laa_multi_step_forms/docs/Setup.md [hunk 2]
```diff
@@ -1 +1,51 @@
...
+
+As gemspec files are not able to point to a github repo bothy the below GEM need
+to be added:
+
+```
+gem 'laa_multi_step_forms', path: 'gems/laa_multi_step_forms'
+gem 'hmcts_common_platform', github: 'ministryofjustice/hmcts_common_platform', tag: 'v0.2.0'
+```
+
+NOTE: hmcts_common_platform provides a list fo Court's and is generica enough it has
```
Comment (willmcb) [inline gems/laa_multi_step_forms/docs/Setup.md:18]: generica --> generic enough to be added to the shared gem.

Comment (willmcb) [reply gems/laa_multi_step_forms/docs/Setup.md:18]: I guess you could add this as a dependency in the gemspec? Rather then ask the use to add it but either is good really

File: gems/laa_multi_step_forms/docs/Setup.md [hunk 3]
```diff
@@ -1 +1,51 @@
...
+been added to teh shared GEM - it may make sense to extract this from the shared
```
Comment (willmcb) [inline gems/laa_multi_step_forms/docs/Setup.md:19]: teh --> the

File: gems/laa_multi_step_forms/docs/Steps.md [hunk 4]
```diff
@@ -1 +1,126 @@
...
+These create a routing record with just a `show` endpoint for the controller
+
+### Edit steps
+
+These create a `edit` and `update` endpoint
+
+## 2. Controllers
+
+By default both edit and show step controllers should inherit from
+`Steps::BaseStepController`. This provides a number of reusable feature the main
```
Comment (willmcb) [inline gems/laa_multi_step_forms/docs/Steps.md:43]: feature --> features, the main one being:

File: gems/laa_multi_step_forms/docs/Steps.md [hunk 5]
```diff
@@ -1 +1,126 @@
...
+one being:
+
+### Navigation stack
+
+By default the navigation stack will automatically be update to include the
```
Comment (willmcb) [inline gems/laa_multi_step_forms/docs/Steps.md:48]: update --> updated

File: gems/laa_multi_step_forms/docs/Steps.md [hunk 6]
```diff
@@ -1 +1,126 @@
...
+This means that routing for multiple paths can be more easily tested outside
+the controller object.
+
+## 3. Value Objects
+
+These are useful when process a list of items - i.e. radio buttons or checkboxes
+where the resulting value/s will be stored in a single field.
+
+This is an alternative to having a lookup table and is preferred due to it NOT
+requiring maintenance in the DB, and being eaier to add/remove values through
```
Comment (willmcb) [inline gems/laa_multi_step_forms/docs/Steps.md:90]: eaier --> easier

File: gems/laa_multi_step_forms/docs/TaskList.md [hunk 7]
```diff
@@ -1 +1,87 @@
-How to create the task list and how it works
\ No newline at end of file
+# Tasklist
+
+This is a generic component for an mutli step form submission process,
```
Comment (willmcb) [inline gems/laa_multi_step_forms/docs/TaskList.md:3]: mutli --> multi

File: gems/laa_multi_step_forms/docs/TaskList.md [hunk 8]
```diff
@@ -1 +1,87 @@
...
+with the goal being to show how far through the steps are user is, as
```
Comment (willmcb) [inline gems/laa_multi_step_forms/docs/TaskList.md:4]: are --> our

File: gems/laa_multi_step_forms/docs/TaskList.md [hunk 9]
```diff
@@ -1 +1,87 @@
...
+well as provide a function to allow a user to retrace there steps.
+
+The tasklist if comprised of two parts:
```
Comment (willmcb) [inline gems/laa_multi_step_forms/docs/TaskList.md:7]: if --> is

File: gems/laa_multi_step_forms/docs/TaskList.md [hunk 10]
```diff
@@ -1 +1,87 @@
...
+
+```ruby
+def can_start?
+  fulfilled?(PreviousTask)
+end
+```
+
+### #in_progress?
+
+Idicates the task is currently in progress, it defaults to checking if the
```
Comment (willmcb) [inline gems/laa_multi_step_forms/docs/TaskList.md:59]: Idicates --> indicates

File: gems/laa_multi_step_forms/docs/TaskList.md [hunk 11]
```diff
@@ -1 +1,87 @@
...
+URL is in the navigation stack.
+
+> NOTE: This should not need to be overwritten
+
+### #completed?
+
+Indicates the task is completed, can eitehr be a DB check or for more complex
```
Comment (willmcb) [inline gems/laa_multi_step_forms/docs/TaskList.md:66]: eitehr --> either

File: gems/laa_multi_step_forms/docs/Validations.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
+# Validations
+
+A number of custoim validators have been created to help simplify using the forms.
```
Comment (willmcb) [inline gems/laa_multi_step_forms/docs/Validations.md:3]: custoim --> custom

File: gems/laa_multi_step_forms/docs/Validations.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+In order to use the default value, just pass true into the validator, otherwise you
+can specify an override:
+
+```ruby
+validates :date, multiparam_date: true # use the defaults values
+validates :next_event_date, multiparam_date: { allow_past: false, allow_future: true }
+```
+
+> NOTE: This is designed to be use in conjunction with the `Type::MultiparamDate` type
```
Comment (willmcb) [inline gems/laa_multi_step_forms/docs/Validations.md:24]: 👍

File: gems/laa_multi_step_forms/docs/Validations.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+## Nested Objects
+
+The `NestedValidator` is designed to allow the `govuk_error_summary` method to render
+error messages that occurred on nested objects. This validator will work with a single
+nested object or an array (i.e. has_many) of nested object.
```
Comment (willmcb) [inline gems/laa_multi_step_forms/docs/Validations.md:30]: nested object --> nested objects?

File: gems/laa_multi_step_forms/docs/Validations.md [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+When validating an array of objects a `name` variable is passed into the I18n resolver
+for the error message, by default this is an integer (starting at 1) for the position
+of the object in the array. It is possible to overwrite this value to instead call a
+message on the underlying object.
+
+```ruby
+class MyForm << Steps::BaseFormObject
```
Comment (willmcb) [inline gems/laa_multi_step_forms/docs/Validations.md:38]: single `<` ranther than `<<`

File: gems/laa_multi_step_forms/docs/Validations.md [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+  private
+
+  def nested_object_name(index)
+    return 'main' if index.zero?
+    "record #{index}"
+  end
+end
+```
+
+> NOTE: Validations rae skipped if no object is passed in
```
Comment (willmcb) [inline gems/laa_multi_step_forms/docs/Validations.md:54]: rae --> are

**Review by willmcb — APPROVED** [review]
Hey this looks good to me. I think the explanations are clear, just a few typos to look at.

I'm wondering though - why not add the gem to a separate repo rather than embed it in your apps repo?

**Review by mhunt3142 — APPROVED** [review]

**Pull request merged**